### PR TITLE
[FLINK-1819][core]Allow access to RuntimeContext from Input and Output formats

### PR DESCRIPTION
**Pull request by sachingoel0101**
1. Introduces new Rich Input and Output formats, similar to Rich Functions.
2. Makes all existing input and output formats rich, without any API breaking changes.
3. Provides RuntimeContext to GenericSink and GenericSource operators
4. DataSourceTask and DataSinkTask set DistributedUDFContext to their input and output formats respectively.
5. Complete access to RuntimeContext, including access to accumulators.

### CODE REVIEW

**Comment by mxm** [comment]
Thanks for the pull request!

You did not convert all InputFormats and OutputFormats. There are a few more, e.g. HadoopInputFormatBase or JDBCInputFormat. Other than that, your pull request looks good. I think it is fine to keep the notion of normal and rich, just like we do for the operator user functions. That also means that we do not break the existing interfaces.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/DataSourceTask.java [hunk 1]
```diff
@@ -75,6 +78,13 @@
 	// cancel flag
 	private volatile boolean taskCanceled = false;
 
+	/**
+	 * The accumulator map used in the RuntimeContext.
+	 */
+	protected Map<String, Accumulator<?,?>> accumulatorMap;
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/DataSourceTask.java]: This field should be final. Or you get rid of the field and just access the accumulators via `getEnvironment().getAccumulatorRegistry().getUserMap()`.

**Comment by sachingoel0101** [comment]
Ah yes. I'll update them in a while. There's actually some problem with the unit test I've written too. Travis fails sporadically.

**Comment by sachingoel0101** [comment]
@mxm , I've ported all existing IO formats to Rich IO formats. The unit test is consistently passing now.

I also added a `postConfigure` function which allows user to access runtime context and setup things which use the runtime context, for example, accumulators. Its usage can be found in the unit test I've added. I found it impossible to guarantee the existence of `RuntimeContext` in the `configure` function itself.

**Comment by StephanEwen** [comment]
Would not not be much simpler to add an interface `RuntimeContextFormat` with a method (`setRuntimeContext()`) and call this method on the formats that implement this interface? That would be three changed classes, done.

I am also a bit sceptical about the difference between input format and rich input format. For transformation functions the difference between Function (a lambda) and RichFunction (many mode methods) makes sense. Input formats are per-se pretty rich...

**Comment by StephanEwen** [comment]
What is the `postConfigure` method for? Input and output formats have already the `open()` method, that is called after `configure()`. The `configure()` method itself is somewhat of a baggage from the past, when all code was shipped as classes + configuration, rather than via closures.

**Comment by sachingoel0101** [comment]
@StephanEwen, it would be much simpler to have an interface. But then we leave the part about implementing the `setRuntimeContext` and `getRuntimeContext` to the user. IMO that's not a very good option. So it makes to extend rich formats from abstract Rich formats. Doing that, I couldn't figure out a better solution if we want to make all existing IO formats rich. To sum it up:
1. If we only defined an interface, we leave the implementation of `setter` and `getter` methods to the user.
2. If we write a Rich abstract class, we have to make all existing IO formats rich since the user might want to extend one of these, and of course can't extend two classes.
If there's a way to achieve this, let me know. I might have missed something.

As for the `postConfigure` method, it was to guarantee that the `RuntimeContext` will be available now. I tried initializing the context in `DataSourceTask` and `DataSinkTask` before calling `configure`, but it seems the format is re-initialized somewhere as that information is lost. I could remove this function and we can just initialize `counters` etc. in `open` call.

**Comment by fhueske** [comment]
Hi @sachingoel0101,

thanks for opening this PR! 
I see your point that users would have to implement `setRuntimeContext()` and `getRuntimeContext()` if we would use an interface. However, I am not sure that this would affect many people because we could implement the methods in all our base Input/OutputFormats which should cover a large portion of new formats. Also, these methods can be implemented in less than a minute. I think @StephanEwen has a point that all Input/OutputFormats are always quite rich. I would opt for the interface solution but implement it for all relevant Input/OutputFormats that Flink provides.

I would also remove the `postConfigure()` method in favor of a clear documentation of the interface stating that the `RuntimeContext`is available in `open()`.

File: flink-tests/src/test/java/org/apache/flink/test/recordJobs/util/DiscardingOutputFormat.java [hunk 2]
```diff
@@ -19,14 +19,14 @@
...
-import org.apache.flink.api.common.io.OutputFormat;
+import org.apache.flink.api.common.io.RichOutputFormat;
 import org.apache.flink.configuration.Configuration;
 import org.apache.flink.types.Record;
 
 /**
  * A simple output format that discards all data by doing nothing.
  */
-public class DiscardingOutputFormat implements OutputFormat<Record> {
+public class DiscardingOutputFormat extends RichOutputFormat<Record> {
```
Comment (fhueske) [inline flink-tests/src/test/java/org/apache/flink/test/recordJobs/util/DiscardingOutputFormat.java]: I don't think the DiscardingOutputFormat needs to access the RuntimeContext

File: flink-core/src/test/java/org/apache/flink/api/common/operators/util/NonRichGenericInputFormat.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+import org.apache.flink.api.common.io.statistics.BaseStatistics;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.core.io.GenericInputSplit;
+
+/**
+ * Generic base class for all inputs that are not based on files.
+ * This is copied from {@link org.apache.flink.api.common.io.GenericInputFormat}
+ * This class doesn't provide access to RuntimeContext.
+ */
+public abstract class NonRichGenericInputFormat<OT> implements InputFormat<OT, GenericInputSplit> {
```
Comment (fhueske) [inline flink-core/src/test/java/org/apache/flink/api/common/operators/util/NonRichGenericInputFormat.java]: I don't think we need this class. Why would somebody need a generic input format without RuntimeContext, if you can have one with RC for free?

File: flink-java/src/main/java/org/apache/flink/api/java/io/DiscardingOutputFormat.java [hunk 4]
```diff
@@ -27,7 +27,7 @@
  *
  * @param <T> The type of the elements accepted by the output format.
  */
-public class DiscardingOutputFormat<T> implements OutputFormat<T> {
+public class DiscardingOutputFormat<T> extends RichOutputFormat<T> {
```
Comment (fhueske) [inline flink-java/src/main/java/org/apache/flink/api/java/io/DiscardingOutputFormat.java]: Doesn't need to access the RuntimeContext.

File: flink-java/src/test/java/org/apache/flink/api/java/type/extractor/TypeExtractorInputFormatsTest.java [hunk 5]
```diff
@@ -130,7 +131,7 @@ public void testQueryableFormatType() {
 	//  Test formats
 	// --------------------------------------------------------------------------------------------
 	
-	public static final class DummyFloatInputFormat implements InputFormat<Float, InputSplit> {
+	public static final class DummyFloatInputFormat extends RichInputFormat<Float, InputSplit> {
```
Comment (fhueske) [inline flink-java/src/test/java/org/apache/flink/api/java/type/extractor/TypeExtractorInputFormatsTest.java]: Test input/output formats don't need access to the RuntimeContext.

File: flink-staging/flink-hbase/src/test/java/org/apache/flink/addons/hbase/example/HBaseWriteStreamExample.java [hunk 6]
```diff
@@ -80,7 +80,7 @@ public void cancel() {
 	 * This class implements an OutputFormat for HBase
 	 *
 	 */
-	private static class HBaseOutputFormat implements OutputFormat<String> {
+	private static class HBaseOutputFormat extends RichOutputFormat<String> {
```
Comment (fhueske) [inline flink-staging/flink-hbase/src/test/java/org/apache/flink/addons/hbase/example/HBaseWriteStreamExample.java]: No need to update this class if the example doesn't use the RuntimeContext.

File: flink-tests/src/test/java/org/apache/flink/test/accumulators/AccumulatorLiveITCase.java [hunk 7]
```diff
@@ -307,7 +307,7 @@ public void flatMap(String value, Collector<Integer> out) throws Exception {
 
 	}
 
-	private static class WaitingOutputFormat implements OutputFormat<Integer> {
+	private static class WaitingOutputFormat extends RichOutputFormat<Integer> {
```
Comment (fhueske) [inline flink-tests/src/test/java/org/apache/flink/test/accumulators/AccumulatorLiveITCase.java]: Test OF does not use the RuntimeContext.

File: flink-core/src/test/java/org/apache/flink/api/common/operators/util/NonRichGenericInputFormat.java (hunk 3 above)
Comment (sachingoel0101) [inline flink-core/src/test/java/org/apache/flink/api/common/operators/util/NonRichGenericInputFormat.java]: This was only for testing purposes. I needed to verify that I didn't break any functionality by making every IO format rich. That is, to verify the class casting behavior in the Pact tasks.

Comment (fhueske) [inline flink-core/src/test/java/org/apache/flink/api/common/operators/util/NonRichGenericInputFormat.java]: Ah, sorry. Didn't notice it is in test scope. Good have it for tests!

**Comment by sachingoel0101** [comment]
@fhueske , what do we settle on then? Should I keep it as abstract class or change it to an interface?
Also, which classes need to be Rich? Can I assume that the ones you didn't comment on are okay?

File: flink-core/src/test/java/org/apache/flink/api/common/operators/util/NonRichGenericInputFormat.java (hunk 3 above)
Comment (sachingoel0101) [inline flink-core/src/test/java/org/apache/flink/api/common/operators/util/NonRichGenericInputFormat.java]: No problem! :)

**Comment by sachingoel0101** [comment]
@fhueske, in the life cycle of an Input format, the `open` method is called several times in each parallel instance of the `DataSource` operator. Consider for example, the simple example of adding an accumulator. Since access to `RuntimeContext` is enabled only after configure, we will end up adding the same accumulator several times, which of course would fail. This is why I decided to add a `postConfigure` method.

**Comment by StephanEwen** [comment]
I am pretty much against `postConfigure()`. There are so many life cycle methods already, we should really not make it more complicated (I think even `configure()` is a problem, but that's an artifact).

It is not so hard to make it right in the `open()` method, and there are methods that getOrCreate accumulators, afaik.

**Comment by sachingoel0101** [comment]
Okay I guess we could leave the proper usage of `RuntimeContext` to the user. We make it explicit that the `RuntimeContext` will not be available in `configure` in the documentation, and if someone is overriding the `open` method, they likely know it'll be called several times and can take care of not performing, for example, duplicate `accumulator` additions. What do you think?

**Comment by sachingoel0101** [comment]
Build failure is on unrelated tests. The recent error on `PartitionedStateCheckpointingITCase` and the other on `KafkaITCase`.

File: flink-core/src/main/java/org/apache/flink/api/common/operators/GenericDataSinkBase.java [hunk 8]
```diff
@@ -328,9 +330,12 @@ public int compare(IN o1, IN o2) {
 		if(format instanceof InitializeOnMaster) {
 			((InitializeOnMaster)format).initializeGlobal(1);
 		}
-		
+
+		if(format instanceof RichOutputFormat){
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/GenericDataSinkBase.java]: Move after `configure()` call

File: flink-core/src/main/java/org/apache/flink/api/common/operators/GenericDataSourceBase.java [hunk 9]
```diff
@@ -200,14 +202,18 @@ public void accept(Visitor<Operator<?>> visitor) {
...
-	
+
 	// --------------------------------------------------------------------------------------------
 	
-	protected List<OUT> executeOnCollections(ExecutionConfig executionConfig) throws Exception {
+	protected List<OUT> executeOnCollections(RuntimeContext ctx, ExecutionConfig executionConfig) throws Exception {
 		@SuppressWarnings("unchecked")
 		InputFormat<OUT, InputSplit> inputFormat = (InputFormat<OUT, InputSplit>) this.formatWrapper.getUserCodeObject();
+
+		if(inputFormat instanceof RichInputFormat){
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/GenericDataSourceBase.java]: Move after `configure()` call

File: flink-core/src/test/java/org/apache/flink/api/common/io/RichInputFormatTest.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+import org.junit.Test;
+
+/**
+ * Tests runtime context access from inside an RichInputFormat class
+ */
+public class RichInputFormatTest {
+
+	@Test
+	public void testCheckRuntimeContextAccess() {
+		final SerializedInputFormat<Record> inputFormat = new SerializedInputFormat<Record>();
```
Comment (fhueske) [inline flink-core/src/test/java/org/apache/flink/api/common/io/RichInputFormatTest.java]: We are trying to reduce the use of the `Record` data type which is an artifact of a deprecated API. Could you replace it by another data type?

File: flink-core/src/test/java/org/apache/flink/api/common/io/RichOutputFormatTest.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+import org.junit.Test;
+
+/**
+ * Tests runtime context access from inside an RichOutputFormat class
+ */
+public class RichOutputFormatTest {
+
+	@Test
+	public void testCheckRuntimeContextAccess() {
+		final SerializedOutputFormat<Record> inputFormat = new SerializedOutputFormat<Record>();
```
Comment (fhueske) [inline flink-core/src/test/java/org/apache/flink/api/common/io/RichOutputFormatTest.java]: Please replace `Record` data type.

File: flink-staging/flink-jdbc/src/main/java/org/apache/flink/api/java/record/io/jdbc/JDBCOutputFormat.java [hunk 12]
```diff
@@ -40,7 +40,7 @@
 import org.apache.flink.types.StringValue;
 import org.apache.flink.types.Value;
 
-public class JDBCOutputFormat implements OutputFormat<Record> {
```
Comment (fhueske) [inline flink-staging/flink-jdbc/src/main/java/org/apache/flink/api/java/record/io/jdbc/JDBCOutputFormat.java]: Class has been removed by a recent commit.

File: flink-tests/src/test/java/org/apache/flink/test/classloading/jar/CustomInputSplitProgram.java [hunk 13]
```diff
@@ -65,7 +65,7 @@ public static void main(String[] args) throws Exception {
 	}
 	// --------------------------------------------------------------------------------------------
 	
-	public static final class CustomInputFormat implements InputFormat<Integer, CustomInputSplit>, ResultTypeQueryable<Integer> {
+	public static final class CustomInputFormat extends RichInputFormat<Integer, CustomInputSplit> implements ResultTypeQueryable<Integer> {
```
Comment (fhueske) [inline flink-tests/src/test/java/org/apache/flink/test/classloading/jar/CustomInputSplitProgram.java]: This is a test class which does not need to be adapted.

File: flink-staging/flink-hadoop-compatibility/src/main/java/org/apache/flink/hadoopcompatibility/mapred/record/HadoopRecordInputFormat.java [hunk 14]
```diff
@@ -37,7 +37,7 @@
 import org.apache.hadoop.mapred.RecordReader;
 import org.apache.hadoop.util.ReflectionUtils;
 
-public class HadoopRecordInputFormat<K, V> implements InputFormat<Record, HadoopInputSplit> {
+public class HadoopRecordInputFormat<K, V> extends RichInputFormat<Record, HadoopInputSplit> {
```
Comment (fhueske) [inline flink-staging/flink-hadoop-compatibility/src/main/java/org/apache/flink/hadoopcompatibility/mapred/record/HadoopRecordInputFormat.java]: Class will be removed soon. Not strictly necessary to adapt it.

File: flink-staging/flink-hadoop-compatibility/src/main/java/org/apache/flink/hadoopcompatibility/mapred/record/HadoopRecordOutputFormat.java [hunk 15]
```diff
@@ -37,7 +37,7 @@
 import org.apache.hadoop.util.ReflectionUtils;
 
 
-public class HadoopRecordOutputFormat<K,V> implements OutputFormat<Record> {
+public class HadoopRecordOutputFormat<K,V> extends RichOutputFormat<Record> {
```
Comment (fhueske) [inline flink-staging/flink-hadoop-compatibility/src/main/java/org/apache/flink/hadoopcompatibility/mapred/record/HadoopRecordOutputFormat.java]: Class will be removed soon. Not strictly necessary to adapt it.

File: flink-core/src/main/java/org/apache/flink/api/common/io/ReplicatingInputFormat.java [hunk 16]
```diff
@@ -59,7 +60,7 @@
  * @see org.apache.flink.api.common.operators.base.FilterOperatorBase
  * @see org.apache.flink.api.common.operators.base.MapPartitionOperatorBase
  */
-public final class ReplicatingInputFormat<OT, S extends InputSplit> implements InputFormat<OT, S> {
+public final class ReplicatingInputFormat<OT, S extends InputSplit> extends RichInputFormat<OT, S> {
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/api/common/io/ReplicatingInputFormat.java]: The `ReplicatingInputFormat` can wrap any kind of `InputFormat` (with or without RuntimeContext). 
Therefore, it needs to extend `RichInputFormat` but overwrite the `setRuntimeContext()` method and forward the context to its wrapped IF, if this is also a `RichInputFormat`.

Please check the updated JavaDocs as well. The wrapped IF does not need to be rich.

**Comment by fhueske** [comment]
Hi @sachingoel0101,
Thanks for the update! The PR looks good in general.
Regarding the interface vs. abstract class issue, I am a bit undecided with a slight preference towards the abstract class. However, I agree with @StephanEwen that `RichInputFormat` and `RichOutputFormat` are not good names. How about `ContextInputFormat` and `ContextOutputFormat`?

Any other opinions on that matter?

**Comment by sachingoel0101** [comment]
I'm inclined towards the Abstract class too. If we're saying something is context aware, we better provide access to it IMO. It doesn't make sense to ask the user to write it themselves and guarantee that we'd just call the relevant methods at runtime.
As for the names, wouldn't it be better to maintain consistency in naming? All the * rich * functions are named `RichFunctions`, which is why I decided to name these `RichFormat`s too. I do agree though. The IO formats are already pretty * rich * in the sense that they allow the user to control everything that happens at runtime.

**Comment by mxm** [comment]
@StephanEwen @fhueske 
The `Rich` prefix might seem a bit odd but it is a naming convention that is consistent with the user-defined functions which have the RuntimeContext available.
As for `abstract classes vs interface` I agree with @sachingoel0101 that it makes sense to implement access to the RuntimeContext once for the user. We may also add other "rich" functions; if we implemented that using an interface we would have to change it later on or add another interface.

I think Input/Ouput formats should be "rich" by default. However, we might not want to break the API for existing external implementations.

**Comment by StephanEwen** [comment]
Okay, so how about calling the abstract base class`AbstractInputFormat` and have it implement the runtime context and leaves the other methods abstract?

**Comment by sachingoel0101** [comment]
Okay. I think maybe calling them `Rich` maybe an overkill. I will change the names to `Abstract`. 
Do we keep the existing formats *non-rich* or would it be okay to make all of them *rich* [which is the case right now]? The problem with not making them *rich* is that then we're limiting any extending classes to never be *rich*.

**Comment by fhueske** [comment]
I'm in favor of making the existing formats "rich".

**Comment by mxm** [comment]
Yes, existing formats should converted to "rich" formats. The name `Abstract..` makes more sense if it becomes the new default way of interfacing with the rich input/output formats. I still think, calling it `Rich...` makes more sense in terms of consistency with the user-defined functions.

**Comment by sachingoel0101** [comment]
Okay. I'll hold off on making any changes right now. We should reach at a consensus first.

**Comment by fhueske** [comment]
@mxm, @sachingoel0101 get your point. Being consistent with the remaining API is a valid argument. I would be OK with calling the abstract classes `RichInputFormat` and `RichOutputFormat`.

Let's wait for at least 24 hours and merge this PR if nobody speaks up.

**Comment by StephanEwen** [comment]
We are back to square one ;-)

  - `Function`: single abstract method
  - `RichFunction` = 5 methods. I see how that gets rich.

  - `InputFormat`: 8 methods
  - `RichInputFormat`: 10 methods.
We could call it `SlightlyMoreRichInputFormat` ;-) I cannot help but find this very confusing.

Why the urgency to stick with the "Rich" prefix?

**Comment by StephanEwen** [comment]
For transformation functions, there is a clear case for "thin" versus "rich", for Java8 lambdas.
Input formats are a different game. They are super rich by default anyways.

**Comment by StephanEwen** [comment]
The suggestion `AbstractInputFormat` was not so bad, in my opinion.
If you want a name that explains what's happening, you can always call it `InputFormatWithContext` ;-)

**Comment by fhueske** [comment]
To keep it consistent with the remaining API. For functions you need to extend a RichFunction if you want to have access to the RuntimeContext. I agree that the name is not perfect (and I think everybody else got your point as well) but I think it's a valid point to aim for consistency.

**Comment by mxm** [comment]
"Rich" does not refer to the number of methods but the fact that it has the RuntimeContext available. All non-rich variants do not get state inserted. This follows a naming convention in Flink. `AbstractInputFormat` might be a more intuitive name for novices but I'm more inclined to naming consistency.

**Comment by StephanEwen** [comment]
Functions also need to extend RichFunction to have access to `open()` and `close()`.
I think the two things a different enough that any strife for "consistency" is actually pretty "random".
If your thoughts currently revolve around the RuntimeContext, it apprears more consistent. If you thoughts are on the life cycle methods, it seems inconsistent. Random.

I think you should go ahead and just call them "Rich". It is just a name, and what matters is that the JavaDocs describe what it actually means...

**Comment by sachingoel0101** [comment]
I agree with Stephan's argument that addition of context to I/O formats is
a very marginal enhancement. He literally stole my words. :')
However, from my perspective, when I first started using flink, Rich meant
runtime context. The idea of open and close wasn't as nearly exciting as
the runtime context.
What if we changed back to the original name mentioned on jira and make it
`ContextAwareInputFormat`? Would everyone be okay with that?

**Comment by sachingoel0101** [comment]
Oh apologies. I  only saw the first comment on the email thread. I guess
it's more or less settled. I'll leave it up to you guys to make a final
decision on this. :')

**Comment by mxm** [comment]
Thank you for your remarks @StephanEwen.

>I think you should go ahead and just call them "Rich". It is just a name, and what matters is that the JavaDocs describe what it actually means...

The JavaDocs in this pull request properly describe what a "RichInputFormat" is. So we are good to go. Let's stick with "Rich" although I understand your concerns.

I'll merge this later if nobody objects in the meantime :)

**Comment by mxm** [comment]
Thanks for the contribution @sachingoel0101!

**Comment by sachingoel0101** [comment]
Sure. :)

**Pull request closed without merging**